### DISCUSSION TITLE
[Security] Question about security exceptions inside providers

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [symfony/symfony#44647] [Q&A] [answered]
Hello, 

I have an issue with the configuration of the error response codes in my API.

I have a Subresource collection provider that can return an `AccessDeniedHttpException` if the user does not comply with certain criteria.

This endpoint has a `security="is_granted('IS_AUTHENTICATED_FULLY')"` annotation.

I would expect my API to return a 401 if I call this endpoint without being authenticated.

However, I get a 403.

I have reproduced this issue in https://github.com/TristanPouliquen/po_security

If you call /api/custom_entities/1, you will get a 403.

If you remove the exception from the `CustomEntityProvider` and put anything else, you will get the expected 401 error code.

What am I doing wrong?

**Comment by norkunas** [comment]
Probably should be: `security="is_granted('IS_AUTHENTICATED_FULLY')"`

**Reply by TristanPouliquen** [reply]
Yeah, typo in my post, I have got 'IS_AUTHENTICATED_FULLY' in my code :-)

**Comment by norkunas** [comment, marked as answer]
api-platform always calls security attribute after calling data provider because in certain scenarios you need to access to the `object` to determine if user could read that endpoint. if you want to avoid this, you can add a rule to firewall

**Reply by TristanPouliquen** [reply]
Indeed,so there'll be no way around it :/ 

I had created an issue to update the definition of ACL (https://github.com/symfony/symfony/issues/44103) to be able to make them match finely with a given firewall definition, I'll try and push it through to implement system-wide ACL then, thanks
